# Pendahuluan

Di sinilah peran sistem rekomendasi yang akan Anda pelajari di modul ini. Ia ada untuk memfilter data yang melimpah menjadi informasi penting dan bermanfaat bagi perusahaan atau organisasi. 

----

Data mungkin mudah dikumpulkan, tetapi mendapatkan data berkualitas baik, bisa jadi sulit. Anda perlu menginvestasikan sebagian besar waktu untuk membersihkan dan memahami bagaimana karakteristik data sebelum menggunakannya. 
---

---
**"Sistem rekomendasi memprediksi rating atau preferensi pengguna terhadap item tertentu. Rekomendasi ini dibuat berdasarkan perilaku pengguna di masa lalu atau perilaku pengguna lainnya. Jadi, sistem ini akan merekomendasikan sesuatu terhadap pengguna berdasarkan data perilaku atau preferensi dari waktu ke waktu". -Frank Kane**

> Hal penting lain yang perlu kita ketahui mengenai sistem rekomendasi adalah, sistem ini tidak merekomendasikan item secara spesifik. **Ia merekomendasikan sejumlah item yang mungkin cocok dengan preferensi pengguna**. Oleh karena itu, pada sistem rekomendasi, keluarannya berupa “top-N” recommendation. Artinya, mesin akan memberikan sejumlah rekomendasi dengan peringkat teratas sesuai preferensi pengguna. Harapannya, pengguna akan memilih rekomendasi yang diberikan dan memperpanjang interaksi pengguna di situs atau aplikasi. Selain itu, diharapkan pengguna juga melakukan transaksi pembelian terhadap item yang direkomendasikan.

---
Tujuan dan fungsi sistem rekomendasi:

- Meningkatkan jumlah item yang terjual
- Menjual item yang beragam
- Meningkatkan kepuasan pengguna
- pemahaman yang lebih baik tentang preferensi pengguna

## Taksonomi Sistem Rekomendasi

Algoritma sistem rekomendasi dibagi mennjadi dua:
- Non-Personalized
> Pada non-personalized recommender system, setiap orang yang berinteraksi dengan sistem akan menerima daftar rekomendasi yang sama. **Sistem ini memberikan rekomendasi pada pengguna tanpa konteks apa pun, baik itu konteks preferensi maupun rating dari pengguna**. Non-personalized recommender system biasanya merupakan **sistem rekomendasi awal** yang dibangun oleh organisasi/bisnis. Hal ini karena ia tidak memerlukan informasi spesifik tentang pengguna. 

- Personalized 
> Sementara itu, teknik rekomendasi dengan personalisasi dapat diklasifikasikan lagi menjadi beberapa kategori berbeda, antara lain **Content-Based Filtering (berbasis konten), Collaborative Filtering (filter kolaboratif), dan Hybrid method**. Pada kategori collaborative filtering, terdapat beberapa teknik, antara lain:
> - Memory based. Contohnya, user based dan item based
> - Model based. Contohnya, matrix factorization dan pendekatan deep learning.

---
### Personalized

- Content-Based Filtering (berbasis konten)
> Ide dari sistem rekomendasi berbasis konten (content-based filtering) adalah merekomendasikan item yang mirip dengan item yang disukai pengguna di masa lalu. Content-based filtering mempelajari profil minat pengguna baru berdasarkan data dari objek yang telah dinilai pengguna. Algoritma ini bekerja dengan menyarankan item serupa yang pernah disukai di masa lalu atau sedang dilihat di masa kini kepada pengguna. **Semakin banyak informasi yang diberikan pengguna, semakin baik akurasi sistem rekomendasi**. Untuk membuat profil pengguna, dua informasi ini penting bagi sistem dengan pendekatan content-based filtering: 
> - Model preferensi pengguna.
> - Riwayat interaksi pengguna dengan sistem rekomendasi. 

- Collaborative Filtering (filter kolaborarif)
> Collaborative filtering bergantung pada pendapat komunitas pengguna. Ia tidak memerlukan atribut untuk setiap itemnya seperti pada sistem berbasis konten. Collaborative filtering dibagi lagi menjadi dua kategori, yaitu: **model based** (metode berbasis model machine learning, yang terdiri dari cluster based algorithm, matrix factorization, dan Deep learning atau neural network) dan **memory based** (metode berbasis memori, yang terdiri dari user-based dan item-based). 

- Hybrid Recommender System
> Semua teknik sistem rekomendasi memiliki kekuatan dan kelemahan. Collaborative Filtering membutuhkan banyak feedback dari pengguna agar sistem berfungsi dengan baik. Sementara itu, content based filtering membutuhkan deskripsi item/fitur yang baik. **Nah, untuk mendapatkan keunggulan dari masing-masing strategi, sistem rekomendasi hibrida menggabungkan dua atau lebih strategi rekomendasi yang berbeda**. Sebagai contoh, sistem rekomendasi hibrida menggabungkan collaborative filtering, content-based filtering, dan strategi lainnya. Sistem rekomendasi hibrida dibagi menjadi beberapa kategori, antara lain:
> - **Mixed**,  Teknik ini menggabungkan beberapa hasil rekomendasi dari berbagai strategi untuk memberikan rekomendasi akhir.
> - **Weighted**, Pada teknik ini, skor komponen rekomendasi dari beberapa strategi digabungkan secara numerik. Penggabungan ini menggunakan teknik weighted linear functions (fungsi linear berbobot) untuk menghitung peringkat item.
> - **Switching**, Switching atau pengalihan adalah teknik rekomendasi yang dapat beralih antara satu strategi ke strategi lainnya, bergantung pada kondisi tertentu.

## Data Untuk Sistem Rekomendasi

Sistem rekomendasi memroses sejumlah informasi dan secara aktif mengumpulkan berbagai jenis data untuk membangun rekomendasi. Data yang dikumpulkan umumnya mengenai item yang direkomendasikan serta pengguna yang akan menerima rekomendasi. Data ini sangat beragam, bergantung pada teknik rekomendasi yang digunakan.

Pada sistem rekomendasi **sederhana**, data yang diperlukan misalnya berupa penilaian (rating) pengguna. Sedangkan, sistem rekomendasi yang **lebih kompleks** membutuhkan data lain seperti deskripsi, hubungan sosial, dan aktivitas pengguna.

Francesco Ricci et al dalam Recommender Systems Handbook mengatakan, **data untuk sistem rekomendasi terbagi dalam tiga jenis kategori**, antara lain: item, pengguna, dan transaksi (hubungan antara pengguna dan item). Mari kita uraikan satu per satu.

1. Item
> Item merupakan objek yang direkomendasikan. Objek ini bisa dibedakan dari kompleksitas dan nilai atau kegunaannya. 
> - Jika item yang dipilih relevan bagi pengguna, manfaat yang diberikan akan menutupi biaya yang dikeluarkan. Sedangkan, jika item tidak relevan dan tidak memberikan manfaat, nilai dan rekomendasi item tersebut bagi pengguna adalah negatif.
> - Sistem rekomendasi dapat menggunakan berbagai properti dan fitur items. Sebagai contoh dalam sistem rekomendasi film, beberapa fiturnya antara lain, genre, aktor, dan sutradara. Ketiga fitur ini dapat digunakan untuk menggambarkan sebuah film dan mempelajari bagaimana suatu item bergantung pada fitur-fiturnya. 

2. Pengguna (User)
> Setiap pengguna tentu memiliki karakteristik dan tujuan yang sangat beragam. Untuk melakukan personalisasi rekomendasi, sistem membutuhkan berbagai informasi mengenai pengguna. **Jenis informasi yang dibutuhkan bergantung pada teknik rekomendasi yang digunakan.** 

3. Transaksi
> Transaksi bisa disebut sebagai interaksi antara pengguna dengan sistem. Ia merupakan data-data yang berguna bagi algoritma pembuat rekomendasi. Contohnya, log yang menyimpan informasi penting selama terjadinya interaksi antara pengguna dengan sistem. Log transaksi dapat berisi item yang diklik, dipilih atau dibeli oleh pengguna serta kueri pengguna (data yang dimasukkan pengguna ke mesin pencari). 

## Similarity Measure
Pengukuran tingkat kesamaan erat kaitannya dengan perhitungan jarak antar item. Secara umum hubungan antara kesamaan dan jarak adalah sebagai berikut:

- Semakin besar jarak, kesamaan semakin kecil (menuju nol).
- Semakin kecil jarak, kesamaan semakin besar (menuju satu).

Anda dapat menghitung kesamaan dengan beberapa cara. Kita akan membahas tiga cara yang umum digunakan dalam menghitung tingkat kesamaan, antara lain: **Jaccard distance, Cosine Similarity, dan Euclidean distance**.

a. Jaccard Distance
> Teknik yang sering disebut Jaccard index atau Jaccard similarity coefficient ini dikenalkan oleh Paul Jaccard. Teknik ini menghitung seberapa dekat jarak antara dua set sampel. Dalam hal ini, set sampel bisa berupa sekumpulan restoran yang disukai pengunjung. Koefisien ini mengukur kesamaan dengan menghitung irisan antara sampel dibagi dengan gabungan sampel. 

b. Cosine Similarity
> Cosine similarity mengukur kesamaan antara dua vektor dan menentukan apakah kedua vektor tersebut menunjuk ke arah yang sama. Ia menghitung sudut cosinus antara dua vektor. Semakin kecil sudut cosinus, semakin besar nilai cosine similarity. 

> Metrik ini sering digunakan untuk mengukur kesamaan dokumen dalam analisis teks. Sebagai contoh, dalam studi kasus ini, cosine similarity digunakan untuk mengukur kesamaan nama restoran dan nama masakan.

c. Euclidean Distance
> Metode Euclidean distance dihitung berdasarkan jarak antara dua titik (item). Item serupa letaknya akan berdekatan satu sama lain jika diplot dalam ruang berdimensi-n. Sehingga, kita dapat menghitung jarak antar item. Berdasarkan jarak tersebut, kita dapat merekomendasikan item kepada pengguna.

# Latihan Membuat Proyek Sistem Rekomendasi

Bayangkan Anda adalah seorang Data Scientist di perusahaan food delivery service. Setelah sekian lama beroperasi, perusahaan Anda berhasil mengumpulkan berbagai informasi mengenai pelanggan, restoran dan masakannya, serta rating yang diberikan pelanggan untuk restoran-restoran tersebut. Seluruh informasi ini terkumpul dalam [restaurant and consumer](https://archive.ics.uci.edu/ml/datasets/Restaurant+%26+consumer+data) dataset. 

### Problem Statement
Sebagai seorang Data Scientist, tentu Anda ingin memanfaatkan data tersebut untuk meningkatkan transaksi di perusahaan. Kembangkan sebuah sistem rekomendasi restoran untuk menjawab permasalahan berikut:

- Berdasarkan data mengenai pengguna, bagaimana membuat sistem rekomendasi yang dipersonalisasi dengan teknik content-based filtering?
- Dengan data rating yang Anda miliki, bagaimana perusahaan dapat merekomendasikan restoran lain yang mungkin disukai dan belum pernah dikunjungi oleh pengguna? 

### Goals
Untuk  menjawab pertanyaan tersebut, buatlah sistem rekomendasi dengan tujuan atau goals sebagai berikut:

- Menghasilkan sejumlah rekomendasi restoran yang dipersonalisasi untuk pengguna dengan teknik content-based filtering.
- Menghasilkan sejumlah rekomendasi restoran yang sesuai dengan preferensi pengguna dan belum pernah dikunjungi sebelumnya dengan teknik collaborative filtering.

### Solution Statement
Nah, untuk mencapai goals yang diinginkan, kita akan membuat sistem rekomendasi dengan alur sebagai berikut.

1. Data Understanding
> Data Understanding adalah tahap awal proyek untuk memahami data yang dimiliki. Dalam kasus ini, kita memiliki 9 file terpisah mengenai restoran, konsumen, dan rating.

2. Univariate Exploratory Data Analysis
> Pada tahap ini, Anda melakukan analisis dan eksplorasi setiap variabel pada data. Jika dibutuhkan, Anda dapat melakukan eksplorasi lebih lanjut mengenai keterkaitan antara satu variabel dengan variabel lainnya.

3. Data Preprocessing
> Ini merupakan tahap persiapan data sebelum data digunakan untuk proses selanjutnya. Pada tahap ini, Anda akan melakukan penggabungan beberapa file sehingga menjadi satu kesatuan file yang utuh dan siap digunakan dalam tahap pemodelan.

4. Data Preparation
> Pada tahap ini, Anda mempersiapkan data dan melakukan beberapa teknik seperti mengatasi missing value dan menyamakan jenis masakan. Pada sistem rekomendasi berbasis konten (content-based filtering) yang akan Anda kembangkan, satu restoran mewakili satu kategori masakan. Oleh karena itu, Anda perlu mengecek ulang dan memastikan setiap restoran memiliki satu kategori masakan.

5. Model Development dengan Content Based Filtering
> Pada tahap inilah Anda mengembangkan sistem rekomendasi dengan teknik content based filtering. Ingatlah, teknik content based filtering akan merekomendasikan item yang mirip dengan item yang disukai pengguna di masa lalu. Pada tahap ini, Anda akan menemukan representasi fitur penting dari setiap kategori masakan dengan tfidf vectorizer dan menghitung tingkat kesamaan dengan cosine similarity. Setelah itu, Anda akan membuat sejumlah rekomendasi restoran untuk pelanggan berdasarkan kesamaan yang telah dihitung sebelumnya.

6. Model Delopment dengan Collaborative Filtering
> Pada tahap ini, sistem merekomendasikan sejumlah restoran berdasarkan rating yang telah diberikan sebelumnya. Dari data rating pengguna, kita akan mengidentifikasi restoran-restoran yang mirip dan belum pernah dikunjungi oleh pengguna untuk direkomendasikan.

Yuk, langsung kenalan dengan dataset yang akan kita gunakan di proyek ini.

## Data Understanding



In [1]:
# download dataset with google colab
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00232/RCdata.zip

--2022-11-09 14:01:04--  https://archive.ics.uci.edu/ml/machine-learning-databases/00232/RCdata.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 39674 (39K) [application/x-httpd-php]
Saving to: ‘RCdata.zip’

RCdata.zip          100%[===================>]  38.74K  --.-KB/s    in 0.1s    

2022-11-09 14:01:05 (299 KB/s) - ‘RCdata.zip’ saved [39674/39674]



Jika berhasil, file RCdata.zip akan masuk ke sistem storage Google Colab. Kemudian, lakukan unzip file dengan menerapkan kode berikut.

In [2]:
# unzip file
!unzip /content/RCdata.zip

Archive:  /content/RCdata.zip
  inflating: chefmozaccepts.csv      
  inflating: chefmozcuisine.csv      
  inflating: chefmozhours4.csv       
  inflating: chefmozparking.csv      
  inflating: geoplaces2.csv          
  inflating: rating_final.csv        
  inflating: README                  
  inflating: usercuisine.csv         
  inflating: userpayment.csv         
  inflating: userprofile.csv         


Kesembilan file tersebut dapat kita kategorikan menjadi 3, yaitu restaurant, consumers, dan user-item-rating.

| Restaurant | Consumers | User-Item-rating |
| -------- | -------- | --------- |
| chefmozaccepts.csv | usercuisine.csv| rating_final.csv | 
| chefmozcuisine.csv | userpayment.csv | |
| chefmozhours4.csv | userprofile.csv | |
| chefmozparking.csv |
| geoplaces2.csv |

Selanjutnya, mari kita baca data-data di atas dengan menggunakan fungsi pandas.read_csv. Implementasikan kode berikut.


In [3]:
#data loaded
import pandas as pd
accepts = pd.read_csv('/content/chefmozaccepts.csv')
cuisine = pd.read_csv('/content/chefmozcuisine.csv')
hours = pd.read_csv('/content/chefmozhours4.csv')
parking = pd.read_csv('/content/chefmozparking.csv')
geo = pd.read_csv('/content/geoplaces2.csv', encoding='ISO-8859-1')
usercuisine = pd.read_csv('/content/usercuisine.csv')
payment = pd.read_csv('/content/userpayment.csv')
profile = pd.read_csv('/content/userprofile.csv')
rating = pd.read_csv('/content/rating_final.csv')

In [4]:
geo.head()

placeID   latitude   longitude  \
0   134999  18.915421  -99.184871   
1   132825  22.147392 -100.983092   
2   135106  22.149709 -100.976093   
3   132667  23.752697  -99.163359   
4   132613  23.752903  -99.165076   

                                      the_geom_meter  \
0  0101000020957F000088568DE356715AC138C0A525FC46...   
1  0101000020957F00001AD016568C4858C1243261274BA5...   
2  0101000020957F0000649D6F21634858C119AE9BF528A3...   
3  0101000020957F00005D67BCDDED8157C1222A2DC8D84D...   
4  0101000020957F00008EBA2D06DC8157C194E03B7B504E...   

                             name                                 address  \
0                 Kiku Cuernavaca                              Revolucion   
1                 puesto de tacos  esquina santos degollado y leon guzman   
2      El Rincón de San Francisco                         Universidad 169   
3  little pizza Emilio Portes Gil                 calle emilio portes gil   
4                   carnitas_mata                  lic. Emilio portes gil   

              city            state country fax  ...            alcohol  \
0       Cuernavaca          Morelos  Mexico   ?  ...  No_Alcohol_Served   
1           s.l.p.           s.l.p.  mexico   ?  ...  No_Alcohol_Served   
2  San Luis Potosi  San Luis Potosi  Mexico   ?  ...          Wine-Beer   
3        victoria        tamaulipas       ?   ?  ...  No_Alcohol_Served   
4         victoria       Tamaulipas  Mexico   ?  ...  No_Alcohol_Served   

  smoking_area dress_code     accessibility   price                    url  \
0         none   informal  no_accessibility  medium  kikucuernavaca.com.mx   
1         none   informal        completely     low                      ?   
2  only at bar   informal         partially  medium                      ?   
3         none   informal        completely     low                      ?   
4    permitted   informal        completely  medium                      ?   

  Rambience franchise    area other_services  
0  familiar         f  closed           none  
1  familiar         f    open           none  
2  familiar         f    open           none  
3  familiar         t  closed           none  
4  familiar         t  closed           none  

[5 rows x 21 columns]

In [5]:
accepts.head()

placeID             Rpayment
0   135110                 cash
1   135110                 VISA
2   135110  MasterCard-Eurocard
3   135110     American_Express
4   135110     bank_debit_cards

In [6]:
cuisine.head()

placeID        Rcuisine
0   135110         Spanish
1   135109         Italian
2   135107  Latin_American
3   135106         Mexican
4   135105       Fast_Food

In [7]:
hours.head()

placeID         hours                  days
0   135111  00:00-23:30;  Mon;Tue;Wed;Thu;Fri;
1   135111  00:00-23:30;                  Sat;
2   135111  00:00-23:30;                  Sun;
3   135110  08:00-19:00;  Mon;Tue;Wed;Thu;Fri;
4   135110  00:00-00:00;                  Sat;

In [8]:
parking.head()

placeID parking_lot
0   135111      public
1   135110        none
2   135109        none
3   135108        none
4   135107        none

In [9]:
usercuisine.head()

userID          Rcuisine
0  U1001          American
1  U1002           Mexican
2  U1003           Mexican
3  U1004            Bakery
4  U1004  Breakfast-Brunch

In [10]:
payment.head()

userID          Upayment
0  U1001              cash
1  U1002              cash
2  U1003              cash
3  U1004              cash
4  U1004  bank_debit_cards

In [11]:
profile.head()

userID   latitude   longitude smoker     drink_level dress_preference  \
0  U1001  22.139997 -100.978803  false      abstemious         informal   
1  U1002  22.150087 -100.983325  false      abstemious         informal   
2  U1003  22.119847 -100.946527  false  social drinker           formal   
3  U1004  18.867000  -99.183000  false      abstemious         informal   
4  U1005  22.183477 -100.959891  false      abstemious    no preference   

  ambience transport marital_status        hijos  birth_year    interest  \
0   family   on foot         single  independent        1989     variety   
1   family    public         single  independent        1990  technology   
2   family    public         single  independent        1989        none   
3   family    public         single  independent        1940     variety   
4   family    public         single  independent        1992        none   

           personality  religion      activity  color  weight  budget  height  
0    thrifty-protector      none       student  black      69  medium    1.77  
1  hunter-ostentatious  Catholic       student    red      40     low    1.87  
2          hard-worker  Catholic       student   blue      60     low    1.69  
3          hard-worker      none  professional  green      44  medium    1.53  
4    thrifty-protector  Catholic       student  black      65  medium    1.69

In [12]:
rating.head()

userID  placeID  rating  food_rating  service_rating
0  U1077   135085       2            2               2
1  U1077   135038       2            2               1
2  U1077   132825       2            2               2
3  U1077   135060       1            2               2
4  U1068   135104       1            1               2

In [13]:
print('Jumlah data pembayaran yang diterima restoran: ', len(accepts.placeID.unique()))
print('Jumlah data masakan pada restoran: ', len(cuisine.placeID.unique()))
print('Jumlah data waktu buka restoran: ', len(hours.placeID.unique()))
print('Jumlah data lokasi restoran: ', len(geo.placeID.unique()))
print('Jumlah data masakan pengguna: ', len(usercuisine.userID.unique()))
print('Jumlah data profil pengguna: ', len(profile.userID.unique()))
print('Jumlah data penilaian yang diberikan pengguna: ', len(rating.userID.unique()))
print('Jumlah data penilaian restoran: ', len(rating.placeID.unique()))

Jumlah data pembayaran yang diterima restoran:  615
Jumlah data masakan pada restoran:  769
Jumlah data waktu buka restoran:  694
Jumlah data lokasi restoran:  130
Jumlah data masakan pengguna:  138
Jumlah data profil pengguna:  138
Jumlah data penilaian yang diberikan pengguna:  138
Jumlah data penilaian restoran:  130


Datanya memang tidak terlalu banyak karena kita hanya mengambil sampel berdasarkan placeID dan UserID. **Namun, kita nanti akan melihat bagaimana data yang sedikit tetapi berkualitas bisa tetap memberikan rekomendasi yang baik**. Selanjutnya, mari kita melakukan eksplorasi terhadap data dengan teknik univariate exploratory data analysis.

## Univariate Exploratory Data Analysis

Variabel-variabel pada Restaurant UCI dataset adalah sebagai berikut:

| Variabel | Keterangan|
|---|---|
|accepts | merupakan jenis pembayaran yang diterima pada restoran tertentu.
|cuisine | merupakan jenis masakan yang disajikan pada restoran.
|hours | merupakan jadwal buka dan tutupnya restoran.
|parking | merupakan ketersediaan tempat parkir pada restoran.
|geo | merupakan letak restoran.
|usercuisine | merupakan jenis masakan dari data pengguna.
|payment | merupakan jenis pembayaran yang dipakai pengguna.
|profile | merupakan data profil pengguna.

Tahap eksplorasi penting untuk memahami variabel-variabel pada data serta korelasi antar variabel. **Pemahaman terhadap variabel pada data dan korelasinya akan membantu kita dalam menentukan pendekatan atau algoritma yang cocok untuk data kita**. 
> Idealnya, kita melakukan eksplorasi data terhadap seluruh variabel. 

Namun, pada proyek ini, kita akan mengeksplorasi beberapa variabel saja, antara lain: *accept, cuisine, profile*, dan *rating*

- Variabel cuisine dan rating akan digunakan pada model rekomendasi kita. -
- Sedangkan, variabel accept dan profile hanya untuk melihat bagaimana profile dan jenis pembayaran yang digunakan oleh pengguna. 

### Accept Variabel


In [14]:
accepts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1314 entries, 0 to 1313
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   placeID   1314 non-null   int64 
 1   Rpayment  1314 non-null   object
dtypes: int64(1), object(1)
memory usage: 20.7+ KB


Berdasarkan output di atas, kita dapat mengetahui bahwa file accept.csv memiliki 1314 entri. Ingatlah juga, accepts termasuk ke dalam kategori data Restaurant. Oleh karena itu, terdapat dua variabel di sini, yaitu placeID dan Rpayment. 
- PlaceID merupakan ID restoran
- Rpayment merupakan jenis pembayaran yang digunakan pada restoran.

Untuk melihat ada berapa banyak entri yang unik berdasarkan placeID dan jenis-jenis pembayaran unik yang diterima, jalankan kode berikut.

In [15]:
print('Banyak data: ', len(accepts.placeID.unique()))
print('Jenis pembayaran yang diterima: ', accepts.Rpayment.unique())

Banyak data:  615
Jenis pembayaran yang diterima:  ['cash' 'VISA' 'MasterCard-Eurocard' 'American_Express' 'bank_debit_cards'
 'checks' 'Discover' 'Carte_Blanche' 'Diners_Club' 'Visa'
 'Japan_Credit_Bureau' 'gift_certificates']


Terdapat 615 data restoran yang unik dengan 12 jenis pembayaran. Perhatikanlah, dari sebelumnya 1314 entri, **setelah difilter dengan fungsi unique() datanya menjadi sejumlah 615.**

### Cuisine Variabel
Selanjutnya, mari kita eksplorasi variabel cuisine, yaitu jenis masakan yang tersedia di restoran. Mari kita lihat info variabel cuisine dengan menggunakan fungsi info(). 

In [16]:
cuisine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 916 entries, 0 to 915
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   placeID   916 non-null    int64 
 1   Rcuisine  916 non-null    object
dtypes: int64(1), object(1)
memory usage: 14.4+ KB


Untuk melihat jumlah tipe masakan (cuisine) dan nama-nama masakannya, mari kita kembali gunakan fungsi unique().

In [17]:
print('Banyak tipe masakan: ', len(cuisine.Rcuisine.unique()))
print('Tipe masakan: ', cuisine.Rcuisine.unique())

Banyak tipe masakan:  59
Tipe masakan:  ['Spanish' 'Italian' 'Latin_American' 'Mexican' 'Fast_Food' 'Burgers'
 'Dessert-Ice_Cream' 'Hot_Dogs' 'Steaks' 'Asian' 'International'
 'Mongolian' 'Vegetarian' 'Brazilian' 'Cafe-Coffee_Shop' 'Cafeteria'
 'Contemporary' 'Deli-Sandwiches' 'Diner' 'Japanese' 'Sushi' 'Seafood'
 'Chinese' 'Bar' 'Bar_Pub_Brewery' 'Pizzeria' 'Mediterranean' 'American'
 'Family' 'Caribbean' 'African' 'Breakfast-Brunch' 'Regional' 'Afghan'
 'Bakery' 'Game' 'Armenian' 'Vietnamese' 'Korean' 'Thai' 'Barbecue'
 'Polish' 'Dutch-Belgian' 'French' 'German' 'Southwestern' 'Persian'
 'Ethiopian' 'Juice' 'Soup' 'Continental-European' 'Greek' 'Southern'
 'Eastern_European' 'California' 'Bagels' 'Turkish' 'Organic-Healthy'
 'Fine_Dining']


Terdapat 59 tipe masakan yang berbeda dengan nama masakan seperti terlihat pada output kode. 

Perhatikanlah, jenis-jenis masakan di atas. Pada modul selanjutnya kita akan gunakan data cuisine ini untuk memprediksi top-N rekomendasi bagi pengguna. 

### Profile
Seperti yang telah dibahas sebelumnya, profile pengguna kadang diperlukan untuk memahami pola preferensi terhadap suatu item. Pertama, kita lihat terlebih dahulu berapa jumlah kolom dan baris pada variabel profile. 

In [18]:
profile.shape

(138, 19)

Selanjutnya, mari kita lihat fitur apa saja yang terdapat pada variabel profile. Implementasikan kode berikut.

In [19]:
profile.head()

userID   latitude   longitude smoker     drink_level dress_preference  \
0  U1001  22.139997 -100.978803  false      abstemious         informal   
1  U1002  22.150087 -100.983325  false      abstemious         informal   
2  U1003  22.119847 -100.946527  false  social drinker           formal   
3  U1004  18.867000  -99.183000  false      abstemious         informal   
4  U1005  22.183477 -100.959891  false      abstemious    no preference   

  ambience transport marital_status        hijos  birth_year    interest  \
0   family   on foot         single  independent        1989     variety   
1   family    public         single  independent        1990  technology   
2   family    public         single  independent        1989        none   
3   family    public         single  independent        1940     variety   
4   family    public         single  independent        1992        none   

           personality  religion      activity  color  weight  budget  height  
0    thrifty-protector      none       student  black      69  medium    1.77  
1  hunter-ostentatious  Catholic       student    red      40     low    1.87  
2          hard-worker  Catholic       student   blue      60     low    1.69  
3          hard-worker      none  professional  green      44  medium    1.53  
4    thrifty-protector  Catholic       student  black      65  medium    1.69

*Menarik!*. Kita mendapatkan berbagai data mulai dari marital_status, birth_year, smoker, religion, budget, dll. Jika kita ingin membangun sistem rekomendasi yang lebih kompleks, mungkin kita perlu menyertakan fitur-fitur di atas pada model. Data profile ini berguna jika kita ingin membuat sistem rekomendasi berdasarkan demografi atau kondisi sosial pengguna. Namun, untuk studi kasus kali ini, kita tidak akan menggunakan data profile pada model.

Pada latihan dengan content-based filtering nanti, data yang dibutuhkan adalah nama restoran dan cuisine. Kita akan menghitung kesamaan (similarity) cuisine dan nama restoran kemudian membuat rekomendasi berdasarkan kesamaan ini. **Data-data ini dibutuhkan saat kita membuat sistem rekomendasi dengan teknik lain, misalnya, teknik yang melibatkan demografi**.

### Rating


In [20]:
rating.head()

userID  placeID  rating  food_rating  service_rating
0  U1077   135085       2            2               2
1  U1077   135038       2            2               1
2  U1077   132825       2            2               2
3  U1077   135060       1            2               2
4  U1068   135104       1            1               2

Dari fungsi rating.head(), kita dapat mengetahui bahwa data rating terdiri dari 5 kolom dengan tiga kategori rating. Kolom-kolom tersebut antara lain:

- userID, merupakan identitas pengguna.
- placeID, merupakan identitas restoran.
- Rating, merupakan data rating untuk restoran.
- Food_rating, merupakan data rating untuk makanan atau masakan di restoran tersebut.
- Service_rating, merupakan data layanan restoran tersebut.

Nah, untuk melihat distribusi rating pada data, gunakan fungsi describe() dengan menerapkan kode berikut:

In [21]:
rating.describe()

placeID       rating  food_rating  service_rating
count    1161.000000  1161.000000  1161.000000     1161.000000
mean   134192.041344     1.199828     1.215332        1.090439
std      1100.916275     0.773282     0.792294        0.790844
min    132560.000000     0.000000     0.000000        0.000000
25%    132856.000000     1.000000     1.000000        0.000000
50%    135030.000000     1.000000     1.000000        1.000000
75%    135059.000000     2.000000     2.000000        2.000000
max    135109.000000     2.000000     2.000000        2.000000

Dari output di atas, diketahui bahwa nilai maksimum rating adalah 2 dan nilai minimumnya adalah 0. **Artinya, skala rating berkisar antara 0 hingga 2.** 

Untuk melihat berapa pengguna yang memberikan rating, jumlah restoran, dan jumlah rating, jalankan kode berikut.

In [22]:
print('Jumlah userID: ', len(rating.userID.unique()))
print('Jumlah placeID: ', len(rating.placeID.unique()))
print('Jumlah data rating: ', len(rating))

Jumlah userID:  138
Jumlah placeID:  130
Jumlah data rating:  1161


## Data Preprocessing

### Menggabungkan Restoran
Pertama, mari kita identifikasi berapa jumlah seluruh restoran pada dataset. Di sini, kita menggunakan library numpy dan fungsi *concatenate* untuk menggabungkan beberapa file. Pada materi sebelumnya, Anda mengetahui bahwa kesembilan file data kita dibagi ke dalam tiga kategori, yaitu: Restaurant, User, dan Rating. Nah, pada tahap ini, kita akan menggabungkan seluruh data pada kategori Restaurant. Sehingga, kita akan menggunakan placeID yang unik sebagai acuan dalam penggabungan ini.


In [23]:
import numpy as np
 
# Menggabungkan seluruh placeID pada kategori Restaurant
resto_all= np.concatenate((
    accepts.placeID.unique(),
    cuisine.placeID.unique(),
    hours.placeID.unique(),
    parking.placeID.unique(),
    geo.placeID.unique()
))
 
# Mengurutkan data dan menghapus data yang sama
resto_all = np.sort(np.unique(resto_all))
 
print('Jumlah seluruh data restoran berdasarkan placeID: ', len(resto_all))

Jumlah seluruh data restoran berdasarkan placeID:  938


Perhatikanlah, setelah file pada kategori Restaurant digabung, kita memiliki 938 restoran yang unik.

### Menggabungkan Seluruh User

Mari terapkan cara yang sama pada kategori variabel User. Dengan fungsi *concatenate* dari library numpy, gabungkan seluruh data pada kategori variabel User.

In [24]:
# Menggabungkan seluruh userID
user_all = np.concatenate((
    usercuisine.userID.unique(),
    payment.userID.unique(),
    profile.userID.unique()
))

 
# Menghapus data yang sama kemudian mengurutkannya
user_all = np.sort(np.unique(user_all))
 
print('Jumlah seluruh user: ', len(user_all))

Jumlah seluruh user:  138


Menurut Anda, apa yang menarik dari sini? Kita hanya memiliki 138 data pengguna dari 938 restoran yang memiliki rating. 

### Mengetahui Jumlah Rating

In [25]:
# Menggabungkan file accepts, geo, parking, hours ke dalam dataframe resto_info 
resto_info = pd.concat([accepts, geo, parking, hours])

# Menggabungkan dataframe rating dengan resto_info berdasarkan nilai placeID
resto = pd.merge(rating, resto_info, on='placeID', how='left')
resto

userID  placeID  rating  food_rating  service_rating Rpayment   latitude  \
0     U1077   135085       2            2               2     cash        NaN   
1     U1077   135085       2            2               2      NaN  22.150802   
2     U1077   135085       2            2               2      NaN        NaN   
3     U1077   135085       2            2               2      NaN        NaN   
4     U1077   135085       2            2               2      NaN        NaN   
...     ...      ...     ...          ...             ...      ...        ...   
8073  U1068   132660       0            0               0      NaN  23.752943   
8074  U1068   132660       0            0               0      NaN        NaN   
8075  U1068   132660       0            0               0      NaN        NaN   
8076  U1068   132660       0            0               0      NaN        NaN   
8077  U1068   132660       0            0               0      NaN        NaN   

       longitude                                     the_geom_meter  \
0            NaN                                                NaN   
1    -100.982680  0101000020957F00009F823DA6094858C18A2D4D37F9A4...   
2            NaN                                                NaN   
3            NaN                                                NaN   
4            NaN                                                NaN   
...          ...                                                ...   
8073  -99.164679  0101000020957F00003D7905C9DC8157C13FCD1AB7334E...   
8074         NaN                                                NaN   
8075         NaN                                                NaN   
8076         NaN                                                NaN   
8077         NaN                                                NaN   

                                       name  ...     accessibility   price  \
0                                       NaN  ...               NaN     NaN   
1                    Tortas Locas Hipocampo  ...  no_accessibility  medium   
2                                       NaN  ...               NaN     NaN   
3                                       NaN  ...               NaN     NaN   
4                                       NaN  ...               NaN     NaN   
...                                     ...  ...               ...     ...   
8073  carnitas mata calle Emilio Portes Gil  ...        completely     low   
8074                                    NaN  ...               NaN     NaN   
8075                                    NaN  ...               NaN     NaN   
8076                                    NaN  ...               NaN     NaN   
8077                                    NaN  ...               NaN     NaN   

      url Rambience franchise    area other_services parking_lot  \
0     NaN       NaN       NaN     NaN            NaN         NaN   
1       ?  familiar         f  closed           none         NaN   
2     NaN       NaN       NaN     NaN            NaN      public   
3     NaN       NaN       NaN     NaN            NaN         NaN   
4     NaN       NaN       NaN     NaN            NaN         NaN   
...   ...       ...       ...     ...            ...         ...   
8073    ?  familiar         f  closed           none         NaN   
8074  NaN       NaN       NaN     NaN            NaN        none   
8075  NaN       NaN       NaN     NaN            NaN         NaN   
8076  NaN       NaN       NaN     NaN            NaN         NaN   
8077  NaN       NaN       NaN     NaN            NaN         NaN   

             hours                  days  
0              NaN                   NaN  
1              NaN                   NaN  
2              NaN                   NaN  
3     00:00-00:00;  Mon;Tue;Wed;Thu;Fri;  
4     00:00-00:00;                  Sat;  
...            ...                   ...  
8073           NaN                   NaN  
8074           NaN                   NaN  
8075  00:00-23:30;  Mon;Tue;Wed;Thu;Fr

Variabel setelah dilakukan merge sangat banyak. Output di atas hanya menampilkan sebagian variabel saja. Perhatikanlah, kita memiliki banyak sekali missing value di sini. Mari kita cek lebih lanjut dengan kode berikut.

In [26]:
# Cek missing value dengan fungsi isnull()
resto.isnull().sum()

userID               0
placeID              0
rating               0
food_rating          0
service_rating       0
Rpayment          5781
latitude          6917
longitude         6917
the_geom_meter    6917
name              6917
address           6917
city              6917
state             6917
country           6917
fax               6917
zip               6917
alcohol           6917
smoking_area      6917
dress_code        6917
accessibility     6917
price             6917
url               6917
Rambience         6917
franchise         6917
area              6917
other_services    6917
parking_lot       6917
hours             4619
days              4619
dtype: int64

Terdapat banyak missing value pada sebagian besar fitur. Hanya fitur userID, placeID, rating, food_rating, dan service_rating saja yang memiliki 0 missing value. **Selanjutnya, mari kita hitung jumlah rating, food_rating, dan service rating berdasarkan place.ID** dengan kode berikut.

In [27]:
# Menghitung jumlah rating, food_rating, dan service kemudian menggabungkannya berdasarkan placeID
resto.groupby('placeID').sum()

rating  food_rating  service_rating    latitude    longitude
placeID                                                              
132560       12           24               6   95.009216  -396.667653
132561       15           20              20   94.907276  -396.506024
132564       25           25              30   94.923698  -396.580739
132572       90           90              84  332.124707 -1514.890677
132583       24           24              30   75.689162  -396.937328
...         ...          ...             ...         ...          ...
135088       36           42              36  113.256068  -595.319338
135104       48           80              48  166.270875  -694.179039
135106       96           96              96  221.497088 -1009.760928
135108       65           65              55  243.498787 -1110.269437
135109       24           30              18   75.687139  -396.941400

[130 rows x 5 columns]

### Menggabungkakn Data dengan Fitur Nama Resto
Pertama, definisikan variabel all_resto_rate dengan variabel rating yang telah kita ketahui sebelumnya.

In [28]:
# Definisikan dataframe rating ke dalam variabel all_resto_rate
all_resto_rate = rating
all_resto_rate

userID  placeID  rating  food_rating  service_rating
0     U1077   135085       2            2               2
1     U1077   135038       2            2               1
2     U1077   132825       2            2               2
3     U1077   135060       1            2               2
4     U1068   135104       1            1               2
...     ...      ...     ...          ...             ...
1156  U1043   132630       1            1               1
1157  U1011   132715       1            1               0
1158  U1068   132733       1            1               0
1159  U1068   132594       1            1               1
1160  U1068   132660       0            0               0

[1161 rows x 5 columns]

Selanjutnya, untuk mengetahui nama restoran dengan placeID tertentu, mari kita gabungkan data geo yang berisikan placeID dan nama resto berdasarkan placeID dan assign ke variabel all_resto_name dengan fungsi *merge* dari library pandas.

In [29]:
# Menggabungkan all resto_rate dengan dataframe geo berdasarkan placeID
all_resto_name = pd.merge(all_resto_rate, geo[['placeID', 'name']], on='placeID', how='left')
 
# Print dataframe all_resto_name
all_resto_name

userID  placeID  rating  food_rating  service_rating  \
0     U1077   135085       2            2               2   
1     U1077   135038       2            2               1   
2     U1077   132825       2            2               2   
3     U1077   135060       1            2               2   
4     U1068   135104       1            1               2   
...     ...      ...     ...          ...             ...   
1156  U1043   132630       1            1               1   
1157  U1011   132715       1            1               0   
1158  U1068   132733       1            1               0   
1159  U1068   132594       1            1               1   
1160  U1068   132660       0            0               0   

                                       name  
0                    Tortas Locas Hipocampo  
1                     Restaurant la Chalita  
2                           puesto de tacos  
3                   Restaurante Marisco Sam  
4                                      vips  
...                                     ...  
1156                             palomo tec  
1157                   tacos de la estacion  
1158                          Little Cesarz  
1159     tacos de barbacoa enfrente del Tec  
1160  carnitas mata calle Emilio Portes Gil  

[1161 rows x 6 columns]

### Menggabungkan Data dengan Fitur Masakan Resto
Langkah selanjutnya adalah menggabungkan variabel all_resto_name yang kita peroleh dari tahapan sebelumnya dengan fitur masakan resto (cuisine). **Tujuannya, agar kita mengetahui masakan yang disediakan oleh restoran**

In [30]:
# Menggabungkan dataframe cuisine dengan all_resto_name dan memasukkannya ke dalam variabel all_resto 
all_resto = pd.merge(all_resto_name, cuisine, on='placeID', how='left')
all_resto

userID  placeID  rating  food_rating  service_rating  \
0     U1077   135085       2            2               2   
1     U1077   135038       2            2               1   
2     U1077   132825       2            2               2   
3     U1077   135060       1            2               2   
4     U1068   135104       1            1               2   
...     ...      ...     ...          ...             ...   
1326  U1043   132630       1            1               1   
1327  U1011   132715       1            1               0   
1328  U1068   132733       1            1               0   
1329  U1068   132594       1            1               1   
1330  U1068   132660       0            0               0   

                                       name   Rcuisine  
0                    Tortas Locas Hipocampo  Fast_Food  
1                     Restaurant la Chalita        NaN  
2                           puesto de tacos    Mexican  
3                   Restaurante Marisco Sam    Seafood  
4                                      vips    Mexican  
...                                     ...        ...  
1326                             palomo tec    Mexican  
1327                   tacos de la estacion    Mexican  
1328                          Little Cesarz   Pizzeria  
1329     tacos de barbacoa enfrente del Tec    Mexican  
1330  carnitas mata calle Emilio Portes Gil        NaN  

[1331 rows x 7 columns]

Inilah data yang akan kita gunakan untuk membuat sistem rekomendasi, yayy!

## Data Preparation

### Mengatasi Misisng Value

In [31]:
all_resto.isnull().sum()

userID              0
placeID             0
rating              0
food_rating         0
service_rating      0
name                0
Rcuisine          288
dtype: int64

Perhatikanlah, terdapat 288 missing value pada fitur ‘Rcuisine’ (kategori masakan). 288 dari 1331 merupakan jumlah yang signifikan. Sebenarnya sayang jika data missing value ini langsung di-drop begitu saja. **Namun, kita tidak bisa mengidentifikasi nama masakan yang tidak memiliki data ‘Rcuisine’ ini termasuk ke dalam kategori masakan mana**. Oleh karena itu, untuk saat ini kita akan drop saja missing value ini. 

In [32]:
# membersihkan missing value dengan fungsi dropna()
all_resto_clean = all_resto.dropna()
all_resto_clean

userID  placeID  rating  food_rating  service_rating  \
0     U1077   135085       2            2               2   
2     U1077   132825       2            2               2   
3     U1077   135060       1            2               2   
4     U1068   135104       1            1               2   
5     U1068   132740       0            0               0   
...     ...      ...     ...          ...             ...   
1325  U1043   132732       1            1               1   
1326  U1043   132630       1            1               1   
1327  U1011   132715       1            1               0   
1328  U1068   132733       1            1               0   
1329  U1068   132594       1            1               1   

                                    name   Rcuisine  
0                 Tortas Locas Hipocampo  Fast_Food  
2                        puesto de tacos    Mexican  
3                Restaurante Marisco Sam    Seafood  
4                                   vips    Mexican  
5          Carreton de Flautas y Migadas    Mexican  
...                                  ...        ...  
1325                  Taqueria EL amigo     Mexican  
1326                          palomo tec    Mexican  
1327                tacos de la estacion    Mexican  
1328                       Little Cesarz   Pizzeria  
1329  tacos de barbacoa enfrente del Tec    Mexican  

[1043 rows x 7 columns]

In [33]:
# Mengecek kembali missing value pada variabel all_resto_clean
all_resto_clean.isnull().sum()

userID            0
placeID           0
rating            0
food_rating       0
service_rating    0
name              0
Rcuisine          0
dtype: int64

### Menyamakan Jenis Makanan

Sebelum masuk tahap akhir (pemodelan), kita perlu menyamakan nama masakan. Kadang, masakan yang sama memiliki nama atau kategori yang berbeda. Jika dibiarkan, hal ini bisa menyebabkan bias pada data. Sebagai contoh, pada data yang digunakan dalam studi kasus ini, Anda akan menemukan nama masakan ‘Game’. Anda tentu bertanya-tanya, jenis makanan apa ini? Nah, Anda akan menemukan jawabannya pada tahapan ini.  

Pertama, mari kita cek ulang data setelah proses cleaning pada tahap sebelumnya. Kita buat variabel baru bernama **fix_resto** untuk menyimpan dataframe.

In [34]:
# Mengurutkan resto berdasarkan PlaceID kemudian memasukkannya ke dalam variabel fix_resto
fix_resto = all_resto_clean.sort_values('placeID', ascending=True)
fix_resto

userID  placeID  rating  food_rating  service_rating  \
1303  U1087   132560       1            2               1   
1288  U1050   132560       0            2               0   
14    U1067   132560       1            0               0   
42    U1082   132560       0            0               0   
1052  U1013   132572       1            1               0   
...     ...      ...     ...          ...             ...   
438   U1024   135106       1            1               1   
178   U1020   135109       2            2               1   
1071  U1041   135109       1            2               1   
99    U1030   135109       0            0               0   
180   U1051   135109       1            1               1   

                            name   Rcuisine  
1303          puesto de gorditas   Regional  
1288          puesto de gorditas   Regional  
14            puesto de gorditas   Regional  
42            puesto de gorditas   Regional  
1052                Cafe Chaires  Cafeteria  
...                          ...        ...  
438   El Rincón de San Francisco    Mexican  
178                    Paniroles    Italian  
1071                   Paniroles    Italian  
99                     Paniroles    Italian  
180                    Paniroles    Italian  

[1043 rows x 7 columns]

Sekarang, kita memiliki 1043 baris data. Untuk mengecek berapa jumlah restoran yang mencakup data tersebut, jalankan kode berikut.

In [35]:
# Mengecek berapa jumlah fix_resto
len(fix_resto.placeID.unique())

95

In [36]:
# Mengecek kategori masakan yang unik
fix_resto.Rcuisine.unique()

array(['Regional', 'Cafeteria', 'American', 'Mexican', 'Fast_Food',
       'Italian', 'Armenian', 'Pizzeria', 'Japanese', 'Vietnamese',
       'Family', 'International', 'Game', 'Burgers', 'Bakery', 'Bar',
       'Breakfast-Brunch', 'Bar_Pub_Brewery', 'Mediterranean',
       'Cafe-Coffee_Shop', 'Contemporary', 'Seafood', 'Chinese'],
      dtype=object)

Perhatikanlah, di antara semua kategori masakan pada data, ada satu yang menarik, yaitu kategori masakan bernama ‘Game’. Hal-hal seperti ini kadang kita temukan pada data. Penyebabnya bisa beragam, misalnya, kesalahan input data atau perbedaan istilah di suatu daerah dengan daerah lainnya. **Jika menemukan data seperti ini, hal yang perlu kita lakukan adalah mengeksplorasi datanya lebih lanjut dan melakukan analisis**. 

Tentu solusi untuk permasalahan seperti ini akan berbeda tergantung kasusnya. Untuk kasus pada proyek ini, kita akan memilih salah satu penamaan kategori saja untuk digunakan pada data.

Nah, Anda tentu penasaran, restoran mana yang memiliki kategori ‘Game’? Mari kita lihat sama-sama.

In [37]:
# Mengecek kategori masakan Game 
fix_resto[fix_resto['Rcuisine'] == 'Game']

userID  placeID  rating  food_rating  service_rating name Rcuisine
781   U1015   132851       1            1               1  KFC     Game
509   U1052   132851       1            0               2  KFC     Game
708   U1008   132851       1            1               1  KFC     Game
770   U1037   132851       2            2               1  KFC     Game
574   U1069   132851       1            0               0  KFC     Game
1188  U1131   132851       2            2               2  KFC     Game
764   U1111   132851       2            1               0  KFC     Game

Ternyata, Game adalah masakan pada nama restoran KFC. Menarik. Pertanyaan selanjutnya, adakah kategori masakan (Rcuisine) lain untuk KFC? Mari kita cek.

In [38]:
# Mengecek kategori masakan pada nama restoran KFC
fix_resto[fix_resto['name'] == 'KFC']

userID  placeID  rating  food_rating  service_rating name  Rcuisine
781   U1015   132851       1            1               1  KFC      Game
508   U1052   132851       1            0               2  KFC  American
780   U1015   132851       1            1               1  KFC  American
509   U1052   132851       1            0               2  KFC      Game
708   U1008   132851       1            1               1  KFC      Game
707   U1008   132851       1            1               1  KFC  American
770   U1037   132851       2            2               1  KFC      Game
769   U1037   132851       2            2               1  KFC  American
1187  U1131   132851       2            2               2  KFC  American
574   U1069   132851       1            0               0  KFC      Game
1188  U1131   132851       2            2               2  KFC      Game
764   U1111   132851       2            1               0  KFC      Game
763   U1111   132851       2            1               0  KFC  American
573   U1069   132851       1            0               0  KFC  American

Ada! Ternyata, KFC memiliki dua kategori masakan yang berbeda, ‘Game’ dan ‘American’. Tentu ini harus diperbaiki. **Dalam sistem rekomendasi yang akan kita kembangkan, satu restoran memiliki satu kategori masakan**. Dalam hal ini, KFC lebih cocok disebut sebagai restoran dengan kategori masakan American. **Jadi, kita perlu mengganti kategori ‘Game’ dengan ‘American'**. 

In [39]:
# Mengubah nama kategori masakan ‘Game’ menjadi ‘American’
fix_resto = fix_resto.replace('Game', 'American')
fix_resto[fix_resto['name'] == 'KFC']

userID  placeID  rating  food_rating  service_rating name  Rcuisine
781   U1015   132851       1            1               1  KFC  American
508   U1052   132851       1            0               2  KFC  American
780   U1015   132851       1            1               1  KFC  American
509   U1052   132851       1            0               2  KFC  American
708   U1008   132851       1            1               1  KFC  American
707   U1008   132851       1            1               1  KFC  American
770   U1037   132851       2            2               1  KFC  American
769   U1037   132851       2            2               1  KFC  American
1187  U1131   132851       2            2               2  KFC  American
574   U1069   132851       1            0               0  KFC  American
1188  U1131   132851       2            2               2  KFC  American
764   U1111   132851       2            1               0  KFC  American
763   U1111   132851       2            1               0  KFC  American
573   U1069   132851       1            0               0  KFC  American

Dalam sistem rekomendasi yang kita kembangkan, penting untuk memastikan satu restoran mewakili satu kategori masakan. **Tujuannya supaya tidak terjadi dobel atau rangkap kategori dalam satu restoran**. Sehingga, sistem dapat merekomendasikan resto berdasarkan kategori masakannya. 

Berikutnya, kita bisa melanjutkan ke tahap persiapan. Buatlah variabel bernama preparation.

In [40]:
# Membuat variabel preparation yang berisi dataframe fix_resto kemudian mengurutkan berdasarkan placeID
preparation = fix_resto
preparation.sort_values('placeID')

userID  placeID  rating  food_rating  service_rating  \
1303  U1087   132560       1            2               1   
1288  U1050   132560       0            2               0   
14    U1067   132560       1            0               0   
42    U1082   132560       0            0               0   
184   U1055   132572       2            2               2   
...     ...      ...     ...          ...             ...   
1224  U1002   135106       1            1               1   
99    U1030   135109       0            0               0   
178   U1020   135109       2            2               1   
1071  U1041   135109       1            2               1   
180   U1051   135109       1            1               1   

                            name   Rcuisine  
1303          puesto de gorditas   Regional  
1288          puesto de gorditas   Regional  
14            puesto de gorditas   Regional  
42            puesto de gorditas   Regional  
184                 Cafe Chaires  Cafeteria  
...                          ...        ...  
1224  El Rincón de San Francisco    Mexican  
99                     Paniroles    Italian  
178                    Paniroles    Italian  
1071                   Paniroles    Italian  
180                    Paniroles    Italian  

[1043 rows x 7 columns]

Selanjutnya, kita hanya akan menggunakan data unik untuk dimasukkan ke dalam proses pemodelan. **Oleh karena itu, kita perlu menghapus data yang duplikat dengan fungsi drop_duplicates()**. Dalam hal ini, kita membuang data duplikat pada kolom ‘placeID’.

In [41]:
# Membuang data duplikat pada variabel preparation
preparation = preparation.drop_duplicates('placeID')
preparation

userID  placeID  rating  food_rating  service_rating  \
1303  U1087   132560       1            2               1   
1052  U1013   132572       1            1               0   
168   U1118   132583       0            0               0   
24    U1107   132584       2            2               2   
1329  U1068   132594       1            1               1   
...     ...      ...     ...          ...             ...   
681   U1095   135086       1            2               1   
175   U1020   135088       1            2               0   
4     U1068   135104       1            1               2   
488   U1004   135106       2            2               2   
178   U1020   135109       2            2               1   

                                    name   Rcuisine  
1303                  puesto de gorditas   Regional  
1052                        Cafe Chaires  Cafeteria  
168                     McDonalds Centro   American  
24                    Gorditas Dona Tota    Mexican  
1329  tacos de barbacoa enfrente del Tec    Mexican  
...                                  ...        ...  
681          Mcdonalds Parque Tangamanga  Fast_Food  
175                    Cafeteria cenidet  Cafeteria  
4                                   vips    Mexican  
488           El Rincón de San Francisco    Mexican  
178                            Paniroles    Italian  

[95 rows x 7 columns]

**Selanjutnya, kita perlu melakukan konversi data series menjadi list**. Dalam hal ini, kita menggunakan fungsi *tolist()* dari library numpy. Implementasikan kode berikut.

In [42]:
# Mengonversi data series ‘placeID’ menjadi dalam bentuk list
resto_id = preparation['placeID'].tolist()
 
# Mengonversi data series ‘Name’ menjadi dalam bentuk list
resto_name = preparation['name'].tolist()
 
# Mengonversi data series ‘Rcuisine’ menjadi dalam bentuk list
resto_cuisine = preparation['Rcuisine'].tolist()
 
print(len(resto_id))
print(len(resto_name))
print(len(resto_cuisine))

95
95
95


Tahap berikutnya, kita akan membuat dictionary untuk menentukan pasangan key-value pada data resto_id, resto_name, dan resto_cuisine yang telah kita siapkan sebelumnya.

In [43]:
# Membuat dictionary untuk data ‘resto_id’, ‘resto_name’, dan ‘cuisine’
resto_new = pd.DataFrame({
    'id': resto_id,
    'resto_name': resto_name,
    'cuisine': resto_cuisine
})
resto_new

id                          resto_name    cuisine
0   132560                  puesto de gorditas   Regional
1   132572                        Cafe Chaires  Cafeteria
2   132583                    McDonalds Centro   American
3   132584                  Gorditas Dona Tota    Mexican
4   132594  tacos de barbacoa enfrente del Tec    Mexican
..     ...                                 ...        ...
90  135086         Mcdonalds Parque Tangamanga  Fast_Food
91  135088                   Cafeteria cenidet  Cafeteria
92  135104                                vips    Mexican
93  135106          El Rincón de San Francisco    Mexican
94  135109                           Paniroles    Italian

[95 rows x 3 columns]

Selamat! Data kini telah siap untuk dimasukkan ke dalam pemodelan. Yuk, kita lanjut ke materi berikutnya!

## Model Development dengan Content Based Filter

In [44]:
data = resto_new
data.sample(5)

id                resto_name        cuisine
53  134986  Restaurant Las Mananitas  International
10  132663                 tacos abi        Mexican
21  132754           Cabana Huasteca        Mexican
0   132560        puesto de gorditas       Regional
37  132866                   Chaires      Cafeteria

### TFIDF Vectorizer
Teknik ini akan digunakan pada sistem rekomendasi untuk menemukan representasi fitur penting dari setiap kategori masakan.

In [45]:
from sklearn.feature_extraction.text import TfidfVectorizer
 
# Inisialisasi TfidfVectorizer
tf = TfidfVectorizer()
 
# Melakukan perhitungan idf pada data cuisine
tf.fit(data['cuisine']) 
 
# Mapping array dari fitur index integer ke fitur nama
tf.get_feature_names() 

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


['american',
 'armenian',
 'bar',
 'bar_pub_brewery',
 'breakfast',
 'brunch',
 'burgers',
 'cafe',
 'cafeteria',
 'chinese',
 'coffee_shop',
 'contemporary',
 'family',
 'fast_food',
 'international',
 'italian',
 'japanese',
 'mexican',
 'pizzeria',
 'regional',
 'seafood',
 'vietnamese']

In [46]:
# Melakukan fit lalu ditransformasikan ke bentuk matrix
tfidf_matrix = tf.fit_transform(data['cuisine']) 
 
# Melihat ukuran matrix tfidf
tfidf_matrix.shape 

(95, 22)

Perhatikanlah, matriks yang kita miliki berukuran (95, 22). Nilai 95 merupakan ukuran data dan 22 merupakan matrik kategori masakan. 

In [47]:
# Mengubah vektor tf-idf dalam bentuk matriks dengan fungsi todense()
tfidf_matrix.todense()

matrix([[0., 0., 0., ..., 1., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [1., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]])

Selanjutnya, mari kita lihat matriks tf-idf untuk beberapa resto (resto_name) dan kategori masakan (cuisine).

In [48]:
# Membuat dataframe untuk melihat tf-idf matrix
# Kolom diisi dengan jenis masakan
# Baris diisi dengan nama resto

pd.DataFrame(
     tfidf_matrix.todense(),
     columns=tf.get_feature_names(),
     index=data.resto_name,
 ).sample(22, axis=1).sample(10, axis=0)

mexican  italian  cafe  cafeteria  \
resto_name                                                            
Subway                                0.0      0.0   0.0        0.0   
Restaurante El Reyecito               0.0      0.0   0.0        0.0   
La Posada del Virrey                  0.0      0.0   0.0        0.0   
El Herradero Restaurante and Bar      0.0      0.0   0.0        0.0   
Restaurant Bar Coty y Pablo           0.0      0.0   0.0        0.0   
Restaurante y Pescaderia Tampico      0.0      0.0   0.0        0.0   
Restaurante la Cantina                0.0      0.0   0.0        0.0   
cafe punta del cielo                  0.0      0.0   0.0        1.0   
El Oceano Dorado                      1.0      0.0   0.0        0.0   
Hamburguesas saul                     0.0      0.0   0.0        1.0   

                                  bar_pub_brewery  japanese  seafood  burgers  \
resto_name                                                                      
Subway                                        0.0       0.0      0.0      0.0   
Restaurante El Reyecito                       0.0       0.0      0.0      0.0   
La Posada del Virrey                          0.0       0.0      0.0      0.0   
El Herradero Restaurante and Bar              0.0       0.0      0.0      0.0   
Restaurant Bar Coty y Pablo                   0.0       0.0      0.0      0.0   
Restaurante y Pescaderia Tampico              0.0       0.0      1.0      0.0   
Restaurante la Cantina                        1.0       0.0      0.0      0.0   
cafe punta del cielo                          0.0       0.0      0.0      0.0   
El Oceano Dorado                              0.0       0.0      0.0      0.0   
Hamburguesas saul                             0.0       0.0      0.0      0.0   

                                  international  brunch  ...  american  \
resto_name                                               ...             
Subway                                      0.0     0.0  ...       0.0   
Restaurante El Reyecito                     0.0     0.0  ...       0.0   
La Posada del Virrey                        1.0     0.0  ...       0.0   
El Herradero Restaurante and Bar            0.0     0.0  ...       0.0   
Restaurant Bar Coty y Pablo                 0.0     0.0  ...       0.0   
Restaurante y Pescaderia Tampico            0.0     0.0  ...       0.0   
Restaurante la Cantina                      0.0     0.0  ...       0.0   
cafe punta del cielo                        0.0     0.0  ...       0.0   
El Oceano Dorado                            0.0     0.0  ...       0.0   
Hamburguesas saul                           0.0     0.0  ...       0.0   

                                  chinese  armenian  family  contemporary  \
resto_name                                                                  
Subway                                0.0       0.0     0.0           0.0   
Restaurante El Reyecito               0.0       0.0     0.0           0.0   
La Posada del Virrey                  0.0       0.0     0.0           0.0   
El Herradero Restaurante and Bar      0.0       0.0     0.0           0.0   
Restaurant Bar Coty y Pablo           0.0       0.0     0.0           0.0   
Restaurante y Pescaderia Tampico      0.0       0.0     0.0           0.0   
Restaurante la Cantina                0.0       0.0     0.0           0.0   
cafe punta del cielo                  0.0       0.0     0.0           0.0   
El Oceano Dorado                      0.0       0.0     0.0           0.0   
Hamburguesas saul                     0.0       0.0     0.0           0.0   

                                  coffee_shop  vietnamese  breakfast  bar  \
resto_name                                                                  
Subway                                    0.0         0.0        0.0  0.0   
Restaurante El Reyecito                   0.0         0.0        0.0  0.0   
La Posada del Virrey                      0.0         0.0        0.0  0.0   
El Herrad

Output matriks tf-idf di atas menunjukkan Restaurant Bar Fu-hao memiliki kategori bar. Bar Fu-hao, matriks menunjukan bahwa restoran tersebut merupakan resto dengan kategori bar. Hal ini terlihat dari nilai matriks 1.0 pada kategori bar. Selanjutnya, restoran El Rincon de San Francisco termasuk dalam kategori mexican. Sedangkan, restoran Tortas y hamburguesas el gordo termasuk dalam kategori burgers. Demikian seterusnya. 

Sampai di sini, kita telah berhasil mengidentifikasi representasi fitur penting dari setiap kategori masakan dengan fungsi tfidfvectorizer. Kita juga telah menghasilkan matriks yang menunjukkan korelasi antara jenis masakan dengan restoran. Selanjutnya, kita akan menghitung derajat kesamaan antara satu restoran dengan restoran lainnya untuk menghasilkan kandidat restoran yang akan direkomendasikan.

----
### Cosine Similarity



In [49]:
from sklearn.metrics.pairwise import cosine_similarity
 
# Menghitung cosine similarity pada matrix tf-idf
cosine_sim = cosine_similarity(tfidf_matrix) 
cosine_sim

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 1., 0.],
       [0., 0., 0., ..., 1., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

Selanjutnya, mari kita lihat matriks kesamaan setiap resto dengan menampilkan nama restoran dalam 5 sampel kolom (axis = 1) dan 10 sampel baris (axis=0). Jalankan kode berikut.

In [50]:
# Membuat dataframe dari variabel cosine_sim dengan baris dan kolom berupa nama resto
cosine_sim_df = pd.DataFrame(cosine_sim, index= data['resto_name'], columns=data['resto_name'])
print('Shape: ', cosine_sim_df.shape)
 
# Melihat similarity matrix pada setiap resto
cosine_sim_df.sample(5, axis=1).sample(10, axis=0)

Shape:  (95, 95)


resto_name                           McDonalds Centro  la Cantina  \
resto_name                                                          
El cotorreo                                       0.0         0.0   
Restaurant Familiar El Chino                      0.0         0.0   
puesto de gorditas                                0.0         0.0   
Restaurante la Cantina                            0.0         1.0   
Restaurante de Mariscos la Langosta               0.0         0.0   
Little Cesarz                                     0.0         0.0   
Gordas de morales                                 0.0         0.0   
Restaurant de Mariscos de Picon                   0.0         0.0   
Abondance Restaurante Bar                         0.0         0.0   
Subway                                            0.0         0.0   

resto_name                           Mcdonalds Parque Tangamanga  Luna Cafe  \
resto_name                                                                    
El cotorreo                                                  0.0        0.0   
Restaurant Familiar El Chino                                 0.0        0.0   
puesto de gorditas                                           0.0        0.0   
Restaurante la Cantina                                       0.0        0.0   
Restaurante de Mariscos la Langosta                          0.0        0.0   
Little Cesarz                                                0.0        0.0   
Gordas de morales                                            0.0        0.0   
Restaurant de Mariscos de Picon                              0.0        0.0   
Abondance Restaurante Bar                                    0.0        1.0   
Subway                                                       1.0        0.0   

resto_name                           La Cantina Restaurante  
resto_name                                                   
El cotorreo                                             0.0  
Restaurant Familiar El Chino                            0.0  
puesto de gorditas                                      0.0  
Restaurante la Cantina                                  1.0  
Restaurante de Mariscos la Langosta                     0.0  
Little Cesarz                                           0.0  
Gordas de morales                                       0.0  
Restaurant de Mariscos de Picon                         0.0  
Abondance Restaurante Bar                               0.0  
Subway                                                  0.0

Dengan cosine similarity, kita berhasil mengidentifikasi kesamaan antara satu restoran dengan restoran lainnya. Shape (95, 95) merupakan ukuran matriks similarity dari data yang kita miliki. Berdasarkan data yang ada, matriks di atas sebenarnya berukuran 95 restoran x 95 restoran (masing-masing dalam sumbu X dan Y). Artinya, kita mengidentifikasi tingkat kesamaan pada 95 nama restoran. Tapi tentu kita tidak bisa menampilkan semuanya. **Oleh karena itu, kita hanya memilih 10 restoran pada baris vertikal dan 5 restoran pada sumbu horizontal seperti pada contoh di atas**. 

----

### Mendapatkan Rekomendasi

Di sini, kita membuat fungsi resto_recommendations dengan beberapa parameter sebagai berikut:

- Nama_resto : Nama restoran (index kemiripan dataframe).
- Similarity_data : Dataframe mengenai similarity yang telah kita definisikan sebelumnya.
- Items : Nama dan fitur yang digunakan untuk mendefinisikan kemiripan, dalam hal ini adalah ‘resto_name’ dan ‘cuisine’.
- k : Banyak rekomendasi yang ingin diberikan.

Sebelum mulai menulis kodenya, ingatlah kembali definisi sistem rekomendasi yang menyatakan bahwa keluaran sistem ini adalah berupa top-N recommendation. **Oleh karena itu, kita akan memberikan sejumlah rekomendasi restoran pada pengguna yang diatur dalam parameter k**.

In [51]:
def resto_recommendations(nama_resto, similarity_data = cosine_sim_df, items=data[['resto_name', 'cuisine']], k=5):
      """
      Rekomendasi Resto Berdasarkan kemiripan dataframe

      Parameter:
      ~~~
      nama_resto : tipe data string(str)
                            Nama Restoran (index kemiripan dataframe)
      similarity_data : tipe data pd.DataFrame (object)
                             Kesamaan dataframe, simetrik, dengan resto sebagai 
                             indeks dan kolom
        Items : tipe data pd.Dataframe(object)
                      Mengandung kedua nama dan fitur lainnya yang digunakan untuk mendefinisikan kemiripan 
        k : tipe data integer (int)
            Banyaknya jumlah rekomendai yang diberikan
        ~~~

        Pada index ini, kita mengambil k dengan nilai similarity terbesar
        pada index matrix yang diberikan (i).
        """
        # Mengambil data dengan menggunakan argpartition untuk melakukan partisi secara tidak langsung sepanjang sumbu yanng diberikan
        # Dataframe diubah menjadi numpy
        # Range(start, stop, step)
      index = similarity_data.loc[:,nama_resto].to_numpy().argpartition(
            range(-1, -k, -1)
        )
        # Mengambil data dengan similarity terbesar dari index yang ada
      closest = similarity_data.columns[index[-1:-(k+2):-1]]

        # Drop nama_resto agar nama resto yang dicari tidak muncul dalam daftatr rekomendasi
      closest = closest.drop(nama_resto, errors='ignore')
      return pd.DataFrame(closest).merge(items).head(k)

Perhatikanlah, dengan menggunakan **argpartition**, kita mengambil sejumlah nilai k tertinggi dari similarity data (dalam kasus ini: dataframe cosine_sim_df). Kemudian, kita mengambil data dari bobot (tingkat kesamaan) tertinggi ke terendah. Data ini dimasukkan ke dalam variabel closest. Berikutnya, kita perlu menghapus nama_resto yang yang dicari agar tidak muncul dalam daftar rekomendasi. **Dalam kasus ini, nanti kita akan mencari resto yang mirip dengan KFC**, sehingga kita perlu drop nama_resto KFC agar tidak muncul dalam daftar rekomendasi yang diberikan nanti.  

Selanjutnya, mari kita terapkan kode di atas untuk menemukan rekomendasi restoran yang mirip dengan KFC.

In [52]:
# cek cuisine untuk KFC
data[data.resto_name.eq('KFC')]

id resto_name   cuisine
31  132851        KFC  American

Perhatikanlah, KFC masuk dalam kategori cuisine American. Tentu kita berharap rekomendasi yang diberikan adalah resto dengan kategori yang mirip. Nah, sekarang, dapatkan resto recommendation dengan memanggil fungsi yang telah kita definisikan sebelumnya:

In [53]:
# Mendapatkan rekomendasi restoran yang mirip dengan KFC
resto_recommendations('KFC')

resto_name        cuisine
0                VIPS       American
1  tacos los volcanes       American
2     Pizzeria Julios       American
3             Sirlone  International
4    McDonalds Centro       American

Yay! Berhasil! Sistem kita memberikan rekomendasi 5 nama restoran dengan kategori ‘cuisine’ American dan satu kategori International. 